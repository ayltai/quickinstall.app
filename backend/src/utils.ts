import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
