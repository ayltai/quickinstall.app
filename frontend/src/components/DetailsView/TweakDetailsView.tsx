import Divider from '@mui/material/Divider';
import Stack from '@mui/material/Stack';
import React, { type ComponentType, JSX, } from 'react';

import { DynamicTypedInput, } from '../DynamicTypedInput';
import { DetailsView, } from './DetailsView';
import type { DetailsViewProps, } from './DetailsView.types';
import type { TweakDetailsViewProps, } from './TweakDetailsView.types';

const withDynamicTypedInputs = (Component : ComponentType<DetailsViewProps>) => {
    const ComponentWithDynamicTypedInputs : (props : TweakDetailsViewProps) => JSX.Element = (props : TweakDetailsViewProps) => {
        const { parameters, types, values, onChange, ...rest } = props;

        return (
            <Component
                {...rest}
                color='secondary'>
                <Divider />
                <Stack
                    width='100%'
                    spacing={2}>
                    {parameters.map((parameter, index) => {
                        const handleChange = (value? : string | number | boolean) => onChange && onChange(parameter, value);

                        return (
                            <DynamicTypedInput
                                key={parameter}
                                color='secondary'
                                title={parameter}
                                type={types[index]}
                                value={values[index]}
                                onChange={handleChange} />
                        );
                    })}
                </Stack>
            </Component>
        );
    };

    return ComponentWithDynamicTypedInputs;
};

export const TweakDetailsView = withDynamicTypedInputs(DetailsView);
