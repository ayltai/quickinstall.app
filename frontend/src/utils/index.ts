import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
