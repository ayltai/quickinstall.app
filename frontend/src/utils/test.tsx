import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
