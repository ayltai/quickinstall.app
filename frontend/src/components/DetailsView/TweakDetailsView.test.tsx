import React from 'react';

import { render, } from '../../utils/test';
import { TweakDetailsView, } from './TweakDetailsView';

describe('<TweakDetailsView />', () => {
    it('renders correctly', () => expect(render(
        <TweakDetailsView
            description='Description'
            infoUrl='https://example.com'
            source='Homebrew'
            parameters={[
                'Width',
                'Height',
                'Disabled',
                'Colour',
            ]}
            types={[
                'number',
                'number',
                'boolean',
                'string',
            ]}
            values={[
                100,
                200,
                false,
                'red',
            ]} />
    )).toMatchSnapshot());
});
