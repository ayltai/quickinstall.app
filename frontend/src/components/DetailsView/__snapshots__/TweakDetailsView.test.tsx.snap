// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TweakDetailsView /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
          >
            common.description
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            Description
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
          >
            common.website
          </span>
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1g8o8w2-MuiTypography-root-MuiLink-root"
            href="https://example.com"
            target="_blank"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
            >
              https://example.com
            </p>
          </a>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
          >
            common.source
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            Homebrew
          </p>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <div
          class="MuiStack-root css-lbjakd-MuiStack-root"
        >
          <div
            class="MuiStack-root css-crjfvx-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
            >
              Width
            </p>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl css-16bfqta-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  required=""
                  type="text"
                  value="100"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiStack-root css-crjfvx-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
            >
              Height
            </p>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl css-16bfqta-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  required=""
                  type="text"
                  value="200"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiStack-root css-crjfvx-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
            >
              Disabled
            </p>
            <span
              class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
            >
              <span
                class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary css-1kkmc60-MuiButtonBase-root-MuiSwitch-switchBase"
              >
                <input
                  class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                  type="checkbox"
                />
                <span
                  class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
              />
            </span>
          </div>
          <div
            class="MuiStack-root css-crjfvx-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
            >
              Colour
            </p>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl css-16bfqta-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r2:"
                  required=""
                  type="text"
                  value="red"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <span
          class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
        >
          common.description
        </span>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
        >
          Description
        </p>
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <span
          class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
        >
          common.website
        </span>
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1g8o8w2-MuiTypography-root-MuiLink-root"
          href="https://example.com"
          target="_blank"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            https://example.com
          </p>
        </a>
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <span
          class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
        >
          common.source
        </span>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
        >
          Homebrew
        </p>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiStack-root css-lbjakd-MuiStack-root"
      >
        <div
          class="MuiStack-root css-crjfvx-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
          >
            Width
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl css-16bfqta-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                required=""
                type="text"
                value="100"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-crjfvx-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
          >
            Height
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl css-16bfqta-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                required=""
                type="text"
                value="200"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-crjfvx-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
          >
            Disabled
          </p>
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary css-1kkmc60-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
        </div>
        <div
          class="MuiStack-root css-crjfvx-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1lpz5xa-MuiTypography-root"
          >
            Colour
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl css-16bfqta-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r2:"
                required=""
                type="text"
                value="red"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
